RipCurrent/ripcurrent.h

```diff
@@ -0,0 +1,659 @@
+};
+
+template <>
+struct ActualOrDummyEntryTypeImpl<InputPolicy::Accepts> {
+  typedef int type;
+};
+
+template <InputPolicy INPUT>
+using ActualOrDummyEntryType = typename ActualOrDummyEntryTypeImpl<INPUT>::type;
+
+template <InputPolicy INPUT>
+class EntriesConsumer {
+ public:
+  virtual ~EntriesConsumer() = default;
+  virtual void Accept(const ActualOrDummyEntryType<INPUT>&) = 0;
+};
+
+class DummyNoEntryTypeAcceptor : public EntriesConsumer<InputPolicy::DoesNotAccept> {
+ public:
+  virtual ~DummyNoEntryTypeAcceptor() = default;
+  void Accept(const DummyNoEntryType&) override {
+    bricks::Singleton<RipCurrentMockableErrorHandler>().HandleError(
+        "Internal error: Attepmted to send entries into a non-accepting block.");
+  }
+};
+
+// Run context for RipCurrent allows to run it in background, foreground, or scoped.
+// TODO(dkorolev): Only supports `.Sync()` now. Implement everything else.
+class RipCurrentRunContext {
+ public:
+  explicit RipCurrentRunContext(const std::string& error_message)
+      : sync_called_(false), error_message_(error_message) {}
+  RipCurrentRunContext(RipCurrentRunContext&& rhs) : error_message_(rhs.error_message_) {
+    assert(!rhs.sync_called_);
+    rhs.sync_called_ = true;
+  }
+  void Sync() {
+    assert(!sync_called_);
+    sync_called_ = true;
+  }
+  ~RipCurrentRunContext() {
+    if (!sync_called_) {
+      bricks::Singleton<RipCurrentMockableErrorHandler>().HandleError(error_message_);
+    }
+  }
+
+ private:
+  bool sync_called_ = false;
+  std::string error_message_;
+};
+
+template <InputPolicy INPUT, OutputPolicy OUTPUT>
+class InstanceBeingRun : public EntriesConsumer<INPUT> {
+ public:
+  virtual ~InstanceBeingRun() = default;
+};
+
+// Template logic to wrap the above implementations into abstract classes of proper templated types.
+template <InputPolicy INPUT, OutputPolicy OUTPUT>
+class AbstractCurrent : public SharedUniqueDefinition<INPUT, OUTPUT> {
+ public:
+  typedef SharedUniqueDefinition<INPUT, OUTPUT> T_DEFINITION;
+  static constexpr InputPolicy NEXT_INPUT = InputPolicyMatchingOutputPolicy<OUTPUT>::RESULT;
+
+  AbstractCurrent(T_DEFINITION definition) : T_DEFINITION(definition) {}
+  virtual ~AbstractCurrent() = default;
+  virtual std::shared_ptr<InstanceBeingRun<INPUT, OUTPUT>> SpawnAndRun(
+      std::shared_ptr<EntriesConsumer<NEXT_INPUT>>) const = 0;
+
+  struct Traits {
+    constexpr static InputPolicy INPUT_POLICY = INPUT;
+    constexpr static OutputPolicy OUTPUT_POLICY = OUTPUT;
+  };
+
+  Traits UnderlyingType() const;  // Never called, used from `decltype()`.
+};
+
+// The implementation of the "Instantiated building block == shared_ptr<Abstract building block>" paradigm.
+template <InputPolicy INPUT, OutputPolicy OUTPUT>
+class SharedCurrent : public AbstractCurrent<INPUT, OUTPUT> {
+ public:
+  typedef AbstractCurrent<INPUT, OUTPUT> T_BASE;
+  static constexpr InputPolicy NEXT_INPUT = InputPolicyMatchingOutputPolicy<OUTPUT>::RESULT;
```

Comment:
Looks expressive enough to me as is is now. The relation between these classes is already nontrivial to follow, so KISS is my call :)